processor/src/processingManager.js

```diff
@@ -0,0 +1,233 @@
+ *
+ * @param {string} gitDir
+ * @returns
+ */
+async function isRepoReadyForProcessing(gitDir) {
+  const { repoName, ownerName } = parseRepoGitDir(gitDir)
+  const { id, isEmpty, isMirror } = await queryGiteaRepoDetails(ownerName, repoName)
+
+  if (isEmpty) {
+    log.debug(
+      `Repo: ${ownerName}/${repoName} has no branches, it won't be processed!`,
+    )
+    return false
+  }
+
+  if (isMirror) {
+    // See gitea/models/task.go
+    // https://github.com/kitspace/gitea/blob/d1342b36f1b5e4ac0271d0f01c5270cd087f3575/models/task.go#L22
+    const MigrationStatusIsDone = 4
+    const status = await queryGiteaRepoMigrationStatus(id)
+
+    if (status === MigrationStatusIsDone) {
+      log.debug(
+        `Repo: ${ownerName}/${repoName} migration is done, the repo is processable.`,
+      )
+      return true
+    } else {
+      log.debug(
+        `Repo: ${ownerName}/${repoName} is ${status} not ${MigrationStatusIsDone}, the repo is unprocessable.`,
+      )
+      return false
+    }
+  }
+
+  log.debug(`Repo: ${ownerName}/${repoName} is processable.`)
+  return true
+}
+
+/**
+ *
+ * @param {string} gitDir the file-system path for the git repo
+ * @returns {{ownerName: string, repoName: string}} repo details
+ */
+function parseRepoGitDir(gitDir) {
+  const matcher =
+    /^\/gitea-data\/git\/repositories\/(?<ownerName>\w+)\/(?<repoName>\w+)/
+  const { ownerName, repoName } = gitDir.match(matcher)?.groups
+
+  if (ownerName == null || repoName == null) {
+    throw new Error(`Failed to parse gitDir: ${gitDir}`)
+  }
+
+  return { ownerName, repoName }
+}
+
+/**
+ *
+ * @param {string} ownerName
+ * @param {string} repoName
+ * @returns {Promise<{id: string, isEmpty: boolean, isMirror: boolean}>}
+ */
+async function queryGiteaRepoDetails(ownerName, repoName) {
+  const repoQuery = {
+    name: 'fetch-repository',
+    text:
+      'select id, is_mirror, is_empty from repository' +
+      ' where lower(owner_name)=$1 and lower_name=$2',
+    values: [ownerName, repoName],
+  }
+
+  /**
+   * Query the migration task status for a gitea repo, with exponential backoff.
+   * @returns
+   */
+  const queryGiteaRepoWithBackoff = async repoQuery => {
+    const [ownerName, repoName] = repoQuery.values
+
+    const onBackoff = async (num, delay, resolve) => {
+      log.debug(
+        `Backoff started, querying repo ${ownerName}/${repoName}: ${num} ${delay}ms`,
+      )
+      const repoQueryResult = await client.query(repoQuery)
+
+      if (repoQueryResult.rows.length === 1) {
+        const { id, is_mirror, is_empty } = repoQueryResult.rows[0]
+        log.debug(`Repo: ${ownerName}/${repoName} was found`)
+        return resolve({ id, isEmpty: is_empty, isMirror: is_mirror })
+      }
+    }
+
+    const onFail = reject =>
+      reject(
+        new Error(
+          `Repo: ${ownerName}/${repoName} not found after ${MaximumRetries} trials.`,
+        ),
+      )
+
+    return asyncBackoff(onBackoff, onFail, MAXIMUM_NUM_OF_RETRIES)
+  }
```

Comment:
Can this function be defined outside the top level function?

Reply:
I'm wondering about the backoff as well: can we use a Postgres/SQL lib that would take care of the lowere level details?